File: CPP/ByRound/1746/1746_D_Paths_on_the_Tree.cpp
// Problem: CF 1746 D - Paths on the Tree
// https://codeforces.com/contest/1746/problem/D

/*
 * Problem: D. Paths on the Tree
 * 
 * Algorithm/Technique: 
 * - Tree DFS with greedy assignment of paths to maximize score.
 * - For each node in the tree, we compute how many paths go through it.
 * - At each node, we process all children, recursively computing their contribution.
 * - We balance path counts among siblings to satisfy the constraint |c_u - c_v| <= 1.
 * 
 * Time Complexity: O(n log n) per test case due to sorting of child contributions.
 * Space Complexity: O(n) for storing the tree and recursion stack.
 * 
 * The solution uses a depth-first search (DFS) traversal of the tree starting from the root.
 * At every node, we calculate the optimal distribution of paths among its children such that
 * the constraint on adjacent vertices' path counts is satisfied, and the total score is maximized.
 */

#include <algorithm>
#include <functional>
#include <initializer_list>
#include <iostream>
#include <iterator>
#include <string>

using namespace std;
enum { N = 200009 };
basic_string<int> e[N];
int s[N], o;
long long ans;
int dfs(int x, int y) {
  // Add the contribution of the current node to the answer.
  ans += s[x] * 1ll * y;
  // If node is a leaf, no more paths can go through it further.
  if (e[x].empty())
    return s[x];
  // Collect contributions from all children.
  basic_string<int> w;
  for (int o : e[x]) {
    w += dfs(o, y / e[x].size());
  }
  // Sort child contributions in descending order to prioritize higher values.
  sort(begin(w), end(w), greater<int>());
  // Determine how many children will get one extra path.
  int p = y % e[x].size();
  // Add the extra path values to the result.
  for (int i = 0; i < p; ++i)
    ans += w[i];
  // Return the total contribution of this node considering paths going through it.
  return s[x] + w[p];
}
int main() {
  int T, n, i, j, k;
  for (cin >> T; T--;) {
    cin >> n >> o;
    ans = 0;
    // Initialize the adjacency list.
    for (i = 1; i <= n; ++i)
      e[i] = {};
    // Build the tree from parent information.
    for (i = 2; i <= n; ++i)
      cin >> j, e[j] += i;
    // Read scores for each node.
    for (i = 1; i <= n; ++i)
      cin >> s[i];
    // Start DFS from the root node.
    dfs(1, o);
    cout << ans << '\n';
  }
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/D/1746_D_Paths_on_the_Tree.py
# Problem: CF 1746 D - Paths on the Tree
# https://codeforces.com/contest/1746/problem/D

"""
Algorithm: Tree Dynamic Programming with Topological Sort
Time Complexity: O(n)
Space Complexity: O(n)

The problem involves assigning k paths starting from root to maximize a weighted sum,
where each node's contribution is weighted by how many paths cover it.
This is solved using topological sorting and dynamic Programming on the tree.

Key Techniques:
1. Topological sort to process nodes from leaves to root
2. Dynamic Programming to compute the number of paths covering each node
3. Greedy assignment of extra paths to maximize score

Approach:
1. Compute in-degrees and topologically sort nodes
2. For each node in reverse topological order:
   - Compute depth-weighted scores for its children
   - Distribute paths from parent to children
3. For each node in topological order:
   - Assign extra paths to maximize score contribution
   - Accumulate total value

The key insight is to greedily assign extra paths to children with highest scores.
"""

import sys

input = sys.stdin.readline

t = int(input())
for _ in range(t):
    n, k = map(int, input().split())

    p = [0] + list(map(lambda x: int(x) - 1, input().split()))
    s = list(map(int, input().split()))

    # Compute in-degrees for topological sort
    ind = [0] * n
    for v in p[1:]:
        ind[v] += 1

    topo = []
    indc = ind[:]
    st = []
    for i in range(n):
        if ind[i] == 0:
            st.append(i)

    # Build child adjacency list
    ch = [[] for _ in range(n)]
    while st:
        u = st.pop()
        topo.append(u)

        if u == 0:
            continue

        indc[p[u]] -= 1
        if indc[p[u]] == 0:
            st.append(p[u])

        ch[p[u]].append(u)
    assert topo[-1] == 0

    # Compute depth-weighted scores for each node
    dep = [0] * n
    for u in topo[::-1]:
        dep[u] = dep[p[u]] + s[u]

    # Initialize coverage counts and extra assignment tracking
    c = [0] * n
    c[0] = k
    ex = [0] * n  # number of extra paths for each node
    nex = [0] * n  # updated next path count for each node

    out = 0

    # Process nodes in reverse topological order to distribute paths from parent to children
    for u in topo[::-1]:
        if ind[u] == 0:
            continue

        ex[u] = c[u] % ind[u]  # how many children should get one extra path
        drip = c[u] // ind[u]  # how many paths each child gets
        for v in ch[u]:
            c[v] = drip  # distribute paths to each child

    # Final processing to compute total value
    for u in topo:
        if ind[u] == 0:
            out += dep[u] * c[u]
            nex[u] = dep[u]
            continue

        # Get scores of all children
        poss = [nex[v] for v in ch[u]]
        poss.sort(reverse=True)  # sort in descending order for greedy assignment

        # Assign extra paths to maximize value
        out += sum(poss[: ex[u]])  # add contribution of extra paths
        nex[u] = poss[ex[u]]  # remaining maximum value after extra paths

    print(out)


# https://github.com/VaHiX/CodeForces/